Inline Search: Highlight searched term in returned results

interlineae opened the pull request:
## Proposed changes:

This PR aims to increase parity between Inline Search and Instant Search by adding highlighting of the searched term in results. On https://jetpack.com, when searching for a term like `plugin`, results show with the term `plugin` highlighted:

![CleanShot 2025-05-06 at 12 01 11](https://github.com/user-attachments/assets/f9291f01-2bea-4899-aa9b-9c159c810c02)

This is now also the case with Inline Search, as shown with the TwentyTwentyFive theme below. 

![CleanShot 2025-05-06 at 12 03 54](https://github.com/user-attachments/assets/adc8fb26-8970-4833-84f3-4d105eec536d)

In keeping with how Instant Search works, highlighted results are turned in place of post content. On sites with little content this can lead to odd behaviour, as shown below:

> [!NOTE]
> We match for the title, which does not replace the content with the API response.

![Google Chrome 2025-05-08 11 51 12](https://github.com/user-attachments/assets/4ba2fbdc-12ab-4bca-a448-bae3e3aed991)

> [!NOTE]
>  We match for the content, which does replace content with the API response.

![Google Chrome 2025-05-08 11 51 35](https://github.com/user-attachments/assets/0c144762-2c14-4c1e-92cd-8aa3e438f349)

> [!NOTE]
>  We match for the title and content, which does replace content with the API response.

![CleanShot 2025-05-08 at 12 34 45](https://github.com/user-attachments/assets/8f86fc97-0f8f-49c5-9011-f030e975f21c)

**Items to be aware of**

I'm not sure if the content being replaced or not, depending on if content is matched or the title, is jarring. It doesn't seem ideal to truncate content if there are no matches found (just matches for the title).

This implementation plays nice with TwentyTwentyFive's use of the `content` block when returning search results. For other themes that use the `excerpt` block, the API does not replace results. I'm not sure why adding `excerpt` support for the highlighting isn't applying to `excerpt` blocks, @trakos if you had any ideas.

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion

N/A

## Does this pull request change what data or activity we track or use?

N/A

## Testing instructions:

- Either set `add_filter( 'jetpack_search_replace_classic', '__return_true', 20 );` somewhere (e.g. using `Code Snippets` plugin) or add an early `return true` to `should_replace_classic_search()`
- Run a new search. You should expect to see the search term highlighted in the returned results, per above.
- Run searches for other words and ensure word boundaries are respected.

interlineae committed:
Add Search Result Highlighting

- Adds search result highlighting using the same <mark> tag approach as Instant Search.
- Splits up filter__pre_get_posts() with more discrete methods to avoid overloading filter__pre_get_posts() by giving it too much to do.
- Adds new filters and new methods as appropriate.

interlineae committed:
changelog

gibrown commented on projects/packages/search/src/inline-search/class-inline-search.php:
```diff
@@ -421,4 +637,54 @@ public function get_search_result(
+	 * @param string $search_term The search term to highlight.
+	 * @return array Array of patterns to use for highlighting.
+	 */
+	private function prepare_highlight_patterns( $search_term ) {
```
Do we have cases where we really need this?

I think we probably shouldn't try to do this and maintain it in PHP, but rather focus on improving the API if we need something different. Not all languages have word boundaries and there are some cases where regexp matching can be kinda complicated (e.g. special characters and accents).

gibrown commented on projects/packages/search/src/inline-search/class-inline-search.php:
```diff
@@ -301,11 +506,22 @@ public function convert_wp_query_to_api_args( array $args ) {
+		$highlight_options = array(
+			'pre_tags'  => array( '<mark>' ),
+			'post_tags' => array( '</mark>' ),
+			'fields'    => array(
```
Can we add comment highlighting also? One rather nice feature we support is highlighting of comments because we do matching against them. We've also had requests to display highlighting of content in meta, but haven't tried to work on that yet.

Could we inject the highlighted comments into the results similar to what happens in the modal:

<img width="1132" alt="Screenshot 2025-04-16 at 11 23 03" src="https://github.com/user-attachments/assets/a7ff9acf-9d0d-4ca5-b4db-f0d1336ecb1f" />

We probably also don't need to do highlighting on the excerpt. The API can do it, but I don't think we've found a reason for doing highlighting on it in the modal:

https://github.com/Automattic/jetpack/blob/6ccc0b2a85e7b9a3f81a281a502437d09ba0cf36/projects/packages/search/src/instant-search/lib/api.js#L265

gibrown commented on projects/packages/search/src/inline-search/class-inline-search.php:
```diff
@@ -301,11 +506,22 @@ public function convert_wp_query_to_api_args( array $args ) {
+			'fields'    => array(
+				'title'   => array( 'number_of_fragments' => 0 ),
+				'content' => array( 'number_of_fragments' => 0 ),
+				'excerpt' => array( 'number_of_fragments' => 0 ),
```
Right now the API doesn't have a way to customize the fragments in any way. We probably should add that (it has been requested somewhere and is probably more important when displaying in a theme), but maybe in a separate PR unless this becomes a blocker.

gibrown reviewed (commented):
Just a couple of early comments

interlineae committed:
Reorganise class to better group methods

interlineae committed:
First pass at highlighting search corrections

interlineae committed:
Move to a simpler highlighting solution as the previous was overengineered

interlineae committed:
Refactor highlighting handling to better leverage the 1.3api

interlineae committed:
Remove excerpt highlighting as unnecessary at this time

interlineae committed:
Let us move highlighting to a class of its own

interlineae committed:
Merge remote-tracking branch 'origin/trunk' into add/inline-search-term-highlighting

gibrown commented on projects/packages/search/src/inline-search/class-inline-search.php:
```diff
@@ -7,6 +7,8 @@
 
 namespace Automattic\Jetpack\Search;
 
+require_once __DIR__ . '/class-search-highlighter.php';
```
I think there is autoloading. Discussed a bit here: https://github.com/Automattic/jetpack/tree/e6e7ae1698f111ab42ce6b44520871eff04a50ba/projects/packages#package-autoloading

gibrown commented on projects/packages/search/src/inline-search/class-search-highlighter.php:
```diff
@@ -0,0 +1,327 @@
+		}
+
+		// Fallback: Even though the API should provide highlighted titles,
+		// in some cases it doesn't, so we need to apply our own highlighting
```
Do you have cases where this happens? I'd rather we fix it on the API side.

interlineae replied:
Yeah so what's happening right now is that the highlight data from the API doesn't seem to be making it back when requested, despite it being set (via `__MARK__` tag assignment). I'm going to look into this today further as relying on fallbacks isn't ideal, and it might also help with the word boundary concern I raised on Slack.

interlineae replied:
https://github.com/Automattic/jetpack/pull/43110/commits/4b880947987a52ca72cc2cfc32370ef7415419f8

gibrown commented on projects/packages/search/src/inline-search/class-search-highlighter.php:
```diff
@@ -0,0 +1,327 @@
+	public function setup() {
+		add_filter( 'the_title', array( $this, 'filter_highlighted_title' ), 10, 2 );
+		add_filter( 'the_content', array( $this, 'filter_highlighted_content' ), 10, 1 );
+		add_filter( 'comment_text', array( $this, 'filter_highlighted_comment' ), 10, 2 );
```
I think we should actually add any comment matching into the_content. So insert a div with the highlighted content similar to what is done in the popup modal. Most theme search results probably don't display the comments anyways. The point is still to have a relatively compact search result that can indicate what matched (which could be in the content, title, comments, or somewhere else). Usually cases where we don't have highlighting in one of those fields means that another field is matching.

interlineae replied:
https://github.com/Automattic/jetpack/pull/43110/commits/7f45b3877ea87a5e5f1a80f70fa835355887ffbe

interlineae committed:
Merge remote-tracking branch 'origin/trunk' into add/inline-search-term-highlighting

interlineae committed:
We don't need this require

interlineae committed:
Use highlight data returned from the API instead of building our own

interlineae committed:
Address PHPStan

interlineae committed:
Highlight content returned without p tags

When the API sends highlighted data back, it's returned without the expected <p> tags that would wrap content.
Let's add those back.

interlineae committed:
Merge remote-tracking branch 'origin/trunk' into add/inline-search-term-highlighting

interlineae committed:
Cleanup related refactor

interlineae committed:
Merge remote-tracking branch 'origin/trunk' into add/inline-search-term-highlighting

interlineae committed:
Merge remote-tracking branch 'origin/trunk' into add/inline-search-term-highlighting

interlineae committed:
Rename highlighter class to reflect Inline Search

interlineae committed:
Fix tests

interlineae commented on projects/packages/search/tests/php/Inline_Search_Test.php:
```diff
@@ -158,13 +160,13 @@ public static function data_provider(): array {
-					'sort'   => 'date_asc',
-					'langs'  => array( 'en_US' ),
-					'filter' => array(
+					'size'      => '5',
+					'from'      => '0',
```
It looks like `parse_str()` [here](https://github.com/Automattic/jetpack/blame/23eb6faea5ef3e6b52eec43491e17bc79de26b80/projects/packages/search/tests/php/Inline_Search_Test.php#L196) is returning string values, not int, so I've updated this portion of the test in addition to adding the `highlight` array.

interlineae committed:
Fix @var for highlighter to match new class name

trakos commented on projects/packages/search/src/inline-search/class-inline-search.php:
```diff
@@ -314,11 +328,24 @@ public function convert_wp_query_to_api_args( array $args ) {
 		}
 
+		$highlight_fields = array(
+			'post_title',
+			'post_content',
```
Those field names don't work for me when testing, is there some setup I'm missing?

I had to change it to:

```
$highlight_fields = array(
	'title',
	'content',
	'comments',
);
```

to get it to work.

interlineae replied:
This was a reversion. Fixed with https://github.com/Automattic/jetpack/pull/43110/commits/b7a5155f6b56fc7612fac86ff212b56afc2a22ea.

trakos commented on projects/packages/search/src/inline-search/class-inline-search.php:
```diff
@@ -328,6 +355,9 @@ public function convert_wp_query_to_api_args( array $args ) {
 					'must' => $this->build_es_filters( $args ),
 				),
 			),
+			'highlight'    => array(
+				'fields' => $highlight_fields,
+			),
```
This didn't work for me when testing, I had to change it to

```
'highlight_fields' => $highlight_fields,
```

to fix it. I can't figure out how to test it without that change, do you know what I could be doing differently?

interlineae replied:
This was a reversion. Fixed with https://github.com/Automattic/jetpack/pull/43110/commits/b7a5155f6b56fc7612fac86ff212b56afc2a22ea.

trakos commented on projects/packages/search/src/inline-search/class-inline-search.php:
```diff
@@ -314,11 +328,24 @@ public function convert_wp_query_to_api_args( array $args ) {
+		$fields = array(
+			'blog_id',
+			'post_id',
+			'title',
+			'content',
+			'comments',
```
Are those three fields used? I think we're only using `blog_id` and `post_id`, and content/comments/title are fetched from DB anyway to get content in the expected format.

interlineae replied:
That's my understanding based on uggcf://bcratebx.n8p.pbz/fbhepr/kers/jcpbz/choyvp.ncv/erfg/jcpbz-wfba-raqcbvagf/pynff.jcpbz-wfba-ncv-frnepu-fvgr-i1-3-raqcbvag.cuc?e=0op6ons3#103-og.

I had an implementation that used `blog_id` and `post_id`, but it required removing https://github.com/Automattic/jetpack/blob/b7a5155f6b56fc7612fac86ff212b56afc2a22ea/projects/packages/search/src/inline-search/class-inline-search.php#L222-L225, and given the comment I went against that direction. Using `blog_id` and `post_id` did work though. Maybe I've misunderstood?

trakos commented on projects/packages/search/src/inline-search/class-inline-search-highlighter.php:
```diff
@@ -0,0 +1,206 @@
+	 */
+	public function setup() {
+		add_filter( 'the_title', array( $this, 'filter_highlighted_title' ), 10, 2 );
+		add_filter( 'the_content', array( $this, 'filter_highlighted_content' ), 10, 1 );
+		add_filter( 'comment_text', array( $this, 'filter_highlighted_comment' ), 10, 2 );
```
> I'm not sure if the content being replaced or not, depending on if content is matched or the title, is jarring. It doesn't seem ideal to truncate content if there are no matches found (just matches for the title).

I feel like replacing `the_content` and `comment_text` might be undesirable - from what I understand, those are only shown if a site shows full content and comments. In those cases replacing blocks with a highlight, which is a short text fragment (100 characters at most), will be unexpected - especially if some are replaced and some are not.

To me it would make more sense to use highlight in place of excerpts, but leave full content unchanged. What do you think, @anaemnesis @gibrown ?

> For other themes that use the excerpt block, the API does not replace results. 

Replacing excerpt is a bit tricky because `the_excerpt` filter results are truncated, and all HTML is removed from it, so we can't apply highlighting in that way. I've played a bit with some ways to replace excerpts block, and overwriting the whole render worked in my simple test:

```
add_filter( 'render_block_core/post-excerpt', array( $this, 'filter_render_excerpt_block' ), 30, 3 );
```

```
public function filter_render_excerpt_block( $block_content, $block, $instance ) {
	if ( ! isset( $instance->context['postId'] ) || ! $this->is_search_result( $instance->context['postId'] ) ) {
		return $block_content;
	}
	$highlighted_content = $this->get_highlighted_content( $instance->context['postId'] );
	if ( empty( $highlighted_content['content'] ) && empty( $highlighted_content['comments'] ) ) {
		return $block_content;
	}
	if ( ! empty( $highlighted_content['content'] ) ) {
		$block_content = wpautop( $highlighted_content['content'] );
	}
	if ( ! empty( $highlighted_content['comments'] ) ) {
		$block_content .= ' ... ' . $highlighted_content['comments'];
	}
	$more_text           = ! empty( $block['attrs']['moreText'] ) ? '<a class="wp-block-post-excerpt__more-link" href="' . esc_url( get_the_permalink( $block->context['postId'] ) ) . '">' . wp_kses_post( $block['attrs']['moreText'] ) . '</a>' : '';
	$classes = array();
	if ( isset( $block['attrs']['textAlign'] ) ) {
		$classes[] = 'has-text-align-' . $block['attrs']['textAlign'];
	}
	if ( isset( $block['attrs']['style']['elements']['link']['color']['text'] ) ) {
		$classes[] = 'has-link-color';
	}
	$wrapper_attributes    = get_block_wrapper_attributes( array( 'class' => implode( ' ', $classes ) ) );
	$block_content         = '<p class="wp-block-post-excerpt__excerpt">' . $block_content;
	$show_more_on_new_line = ! isset( $block['attrs']['showMoreOnNewLine'] ) || $block['attrs']['showMoreOnNewLine'];
	if ( $show_more_on_new_line && ! empty( $more_text ) ) {
		$block_content .= '</p><p class="wp-block-post-excerpt__more-text">' . $more_text . '</p>';
	} else {
		$block_content .= " $more_text</p>";
	}

	return sprintf( '<div %1$s>%2$s</div>', $wrapper_attributes, $block_content );
}
```

This basically replaces the whole logic for rendering the excerpt block, based on the original rendering code. It might require a bit more testing and making sure it covers all cases.

Alternatively, we could introduce a new block that would have to be used for highlights in search results template in place of content/excerpt, though most users probably wouldn't find that since it would require going to site editor.

interlineae replied:
> I feel like replacing the_content and comment_text might be undesirable - from what I understand, those are only shown if a site shows full content and comments. In those cases replacing blocks with a highlight, which is a short text fragment (100 characters at most), will be unexpected - especially if some are replaced and some are not.

Does this mean we'd prefer a return to my "fallback" handling of highlighting? In that case, we weren't replacing `the_content` or `comment_text` with the API response fragment? This was changed in https://github.com/Automattic/jetpack/pull/43110/commits/4b880947987a52ca72cc2cfc32370ef7415419f8.

Or do you mean more the approach you suggest here, which could be used for other blocks as needed? I like the idea of this approach.

gibrown replied:
Oh... yes this code @trakos made I think matches what I am thinking. We should be using the highlight to replace the text of the content we display. Make it look like what the front end modal does. Whether that means we are filtering the excerpt, content, or whatever the goal is to display highlighted results about what matched and that can really only come from the API.

interlineae replied:
@trakos @gibrown I've made a couple of commits:

https://github.com/Automattic/jetpack/pull/43110/commits/30ced31207b455873180adec83b22748ea7240ea
https://github.com/Automattic/jetpack/pull/43110/commits/6ce249a12e18d1428ecb325419e0faefbe9c4839

Between these two:

- Highlighting applies to both `content` and `excerpt` output.
- Highlighting is working across Classic and Block themes.
- Some themes have explicit styling to remove the background on `mark` tags, e.g. `Astra`.
- `content` output formatting is retained, albeit I've had to use `preg_match` in a few places and I'm iffy on this. But to highlight the difference, the render used to be

![CleanShot 2025-05-13 at 12 47 45](https://github.com/user-attachments/assets/62cc0296-0ea9-42b8-a5d1-0445e90393b0)

And now it's, which is far better imo, but again, `preg_match`. Any thoughts on how to better handle this? I had looked into PHP's [domXPath Class](https://www.php.net/manual/en/class.domxpath.php), but I think that's me stuck overthinking it.

![CleanShot 2025-05-13 at 11 43 30](https://github.com/user-attachments/assets/f7461314-7add-40c3-b51d-e6e9a8405c36)

I need to iterate on this as only the first result is receiving highlighting, but thoughts are nice so in case I'm getting too far into left field.

trakos reviewed (commented):
Thanks a lot for working on this! Overall, the approach looks solid and makes sense. I did run into a few issues when trying it out and had to make some adjustments during testing. I’ve pointed those out in comments. Please let me know if I missed any required setup.

Also, I left an idea around excerpt highlighting - would love your thoughts on that.

interlineae committed:
Fix reversion to highlighting and associated tests

interlineae committed:
Revise content and excerpt approach

- Add new method to return modified excerpt block w/ highlighting.
- Add new method to return modified content block w/ highlighting.
- Retain filter_* methods for classic theme compatibility.

interlineae committed:
New Content block approach

Attempt a new Content block approach that preserves the original content block render, while applying highlighting. This is the difference between returning the API response as an excerpt when rendering the content block, and actually rendering the content block.

interlineae committed:
Cleanup

interlineae committed:
Iterate on the filter_render_content_block() method

interlineae committed:
Remove unnecessary if/foreach

interlineae committed:
No more implicit nullable parameter

gibrown commented on projects/packages/search/src/inline-search/class-inline-search-highlighter.php:
```diff
@@ -0,0 +1,345 @@
+	 *
+	 * @return array|null The highlighted content array or null if not found.
+	 */
+	public function get_highlighted_content( int $post_id ): ?array {
```
Maybe don't bother with this getter method since it isn't getting called from outside of this class.

interlineae replied:
https://github.com/Automattic/jetpack/pull/43110/commits/1f28038b4a1b700f6d01a550f232f20b1cd43c62

gibrown commented on projects/packages/search/src/inline-search/class-inline-search-highlighter.php:
```diff
@@ -0,0 +1,345 @@
+	 * @return string The filtered block content.
+	 * @since $$next-version$$
+	 */
+	public function filter_render_content_block( string $block_content, array $block, object $instance ): string {
```
Can't excerpt and content just use the same filter function? I think we basically want to display the same results whether the theme is showing content or excerpt. Either way we should just show the highlighting that we get from the api.

interlineae replied:
https://github.com/Automattic/jetpack/pull/43110/commits/dad15d9e733d567f15ebd410e2df68e3c8537add

gibrown commented on projects/packages/search/src/inline-search/class-inline-search-highlighter.php:
```diff
@@ -0,0 +1,345 @@
+	 * @return string The highlighted content.
+	 * @since $$next-version$$
+	 */
+	private function apply_highlighting( string $content, array $terms ): string {
```
We shouldn't need to do any highlighting in php

interlineae replied:
I was speaking with @trakos regarding...

uggcf://tvguho.n8p.pbz/Nhgbznggvp/jcpbz/oybo/56o10647sq4936947n889sq9ssn145o9o1s74055/choyvp.ncv/erfg/jcpbz-wfba-raqcbvagf/pynff.jcpbz-wfba-ncv-frnepu-fvgr-i1-3-raqcbvag.cuc#Y924-og

...to be sure that I understood correctly, and indeed, the API returns ~100 characters (max) fragment when building highlighting and returning a response.

This is why I've chosen to do highlighting in PHP using `apply_highlighting()`. However, after talking with Piotr, I started to think that this PR might benefit from being reoriented towards returning highlighting for uses of the `excerpt` block only, since an excerpt and a ~100-character fragment are consonant with each other.

We could then look at highlighting for search results that use the `content` block in the future, but either way, I'm not sure how we would highlight in the content block, to keep formatting, without some kind of PHP processing after the ES response is received. Thoughts?

**Edit:** Ohh, but you also mean returning the fragment for `content` too, right, per https://github.com/Automattic/jetpack/pull/43110#discussion_r2087278639? Yeah that makes sense.

interlineae committed:
Remove unnecessary getter

interlineae committed:
Remove unneeded method

- Use the 'excerpt' highlighting method for both excerpts and content
- Remove default parameters from two add_filters() since they're the default

interlineae committed:
Rename method

interlineae committed:
We were not actually using

interlineae committed:
Update comment

interlineae committed:
Merge remote-tracking branch 'origin/trunk' into add/inline-search-term-highlighting

interlineae committed:
Update inline filters test

interlineae committed:
Remove content highlighting for excerpt highlighting exclusively

interlineae committed:
Remove comment highlighting as too problematic across themes

interlineae committed:
This method doesn't exist anymore

trakos reviewed (approved):
Thank you for the update!
